devito/arguments.py

```diff
@@ -1,68 +1,72 @@
 import abc
 
 import numpy as np
+from collections import OrderedDict
+from sympy import Symbol
 from cached_property import cached_property
-from collections import namedtuple
+from collections import defaultdict, namedtuple, OrderedDict
+from functools import reduce
+from itertools import chain
 
 from devito.exceptions import InvalidArgument
-from devito.logger import debug
-
-""" This module provides a set of classes that help in processing runtime arguments for
-    kernels generated by devito. There are two class hierarchies here:
-    - ArgumentProvider: These are for objects that might be used in the expression
-      provided to the operator as symbols but might resolve to runtime arguments after
-      code generation. Each ArgumentProvider provides one (or more) Argument
-      object(s).
-    - Argument: Classes inheriting from this are for objects that represent the
-      argument itself. Each ArgumentProvider might provide one or more such objects
-      which are used as placeholders for the argument as well as for verification and
-      derivation of default values.
+from devito.logger import debug, error
+from devito.tools import filter_ordered, flatten, GenericVisitor
+from devito.function import CompositeFunction, SymbolicFunction
+from devito.dimension import Dimension
+from devito.ir.support.stencil import retrieve_offsets
+
+
+""" This module contains a set of classes and functions to deal with runtime arguments
+to Operators. It represents the arguments and their relationships as a DAG (N, E) where
+every node (N) is represented by an object of class Parameter and every edge is an object
+of class Dependency. 
+The various class hierarchies are explained here:
+Parameter:
 """
 
-
-class Argument(object):
-
-    """ Abstract base class for any object that represents a run time argument for
-        generated kernels.
+class Parameter(object):
+    """ Abstract base class for any object that represents a node in the dependency
+        graph. It may or may not represent a runtime argument. 
     """
-
-    __metaclass__ = abc.ABCMeta
-
+    is_Argument = False
```

Comment:
Should we call the class ``Parameter`` or ``Argument``?

I think we should come up with a convention and fix this throughout the whole code base.

If there's a distinction between them, it should be made clear

Reply:
There is a difference. Not all entities that depend on arguments or that arguments depend on are runtime arguments themselves. The most common example are `Dimension`s. While `Dimension`s appear in the dependency tree - the symbolic `Dimension` object for `time`, for example carries the only relationship between the array being worked on (example `u=TimeData(..., save=True)`) and the runtime argument called `time_e` (which is an argument passed to the C function at execution time). `time` itself is never passed to the C function but without it we wouldn't be able to build a relationship between the allocated memory for `u` and the maximum number of allowed timesteps. All nodes in the dependency tree are required to be of type `Parameter`, but that doesn't make them runtime arguments. Runtime arguments are of type `Argument`. In this example, `time` would appear in the dependency tree as a `Parameter` object while `time_e` would be of type `Argument`. 
This is partially explained in the docstring but I will elaborate using this comment there as well.

Reply:
OK, makes sense and thanks for clarifying this in the docs. 

(nitpicking, maybe): I wonder whether using ``RuntimeArgument``, instead of ``Argument`` , would help? The distinction would then become immediately clear

Reply:
Maybe an ASCII diagram of the class hierarchy would help

Reply:
Done